javascript/src/refactor.js

```diff
@@ -0,0 +1,648 @@
+                    }
+                }
+            },
+
+            fallback: 'iteration'
+        });
+
+        return redundantFunctions;
+    }
+
+    // Given a list of functions to be replaced, remove all function declarations
+    // replace all calls to the functions with the corresponding literals
+    pruneFuncReferences(pruneList) {
+        var engine = this;
+
+        estraverse.replace(this.ast, {
+            enter: function (node, parent) {
+                if (node.type === "FunctionDeclaration" && (node.id.name in pruneList)) {
+                    if (engine.keep_comments) {
+                        engine.moveAllCommentsToSiblings(node, parent);
+                    }
+                        
+                    engine.changed = true;
+                    this.remove();
+                } else if (node.type === "VariableDeclarator" && (node.id.name in pruneList)) {
+                    engine.changed = true;
+                    this.remove();
+                } else if (node.type === "CallExpression" && (node.callee.name in pruneList)) {
+                    if (pruneList[node.callee.name]) {
+                        engine.changed = true;
+                        return engine.trueLiteral();
+                    } else {
+                        engine.changed = true;
+                        return engine.falseLiteral();
+                    }
+                }
+            },
+
+            leave: function(node, parent) {
+                if (node.type === "VariableDeclaration" && node.declarations.length === 0) {
+                    engine.preserveCommentsBasedOnOption(node, parent, engine.keep_comments);
+                    engine.changed = true;
+                    this.remove();
+                }
+            },  
+
+            fallback: 'iteration'
+        });
+
+        return engine.changed;
+    }
+
+    // Remove the piranha property from literals so code generator can work properly
+    finalizeLiterals() {
+        var engine = this;
+
+        estraverse.traverse(this.ast, {
+            enter: function (node) {
+                if (engine.isPiranhaLiteral(node)) {
+                    delete node.createdByPiranha;
+                }
+            }
+        });
+    }
+
+    // Calls the entire pipeline
+    refactorPipeline() {
+        var iterations = 0;
+        this.changed = true;
+        
+        this.flagAPIToLiteral();
+    
+        while (this.changed && iterations < this.max_cleanup_steps) {
+            this.changed = false;
+
+            this.evalBoolExpressions();
+            this.reduceIfStatements();
+
+            var redundantVarnames = this.getRedundantVarnames();
+            this.pruneVarReferences(redundantVarnames);
+
+            var functionsWithSingleReturn = this.getFunctionsWithSingleReturn();
+            var redundantFunctions = this.getRedundantFunctions(functionsWithSingleReturn)
+            this.pruneFuncReferences(redundantFunctions);
```

Comment:
So, perhaps a naive question (remember, I don't do much JS coding) but: this version of Piranha operates on one `*.js` file at the time, so, what happens with exported functions that are now reduced to a constant but called from other JS files? Do we do deep clean across files? And if not (which is totally fine!), do we have any check to prevent functions used by other JS files from being removed?

Reply:
Not a naive question at all. In its current form, PiranhaJS ignores `import` statements and refactors each source file independently. That said, refactoring across files is one of several enhancements we plan to introduce in the future.

Reply:
Actually, it's fine to leave it like that for this PR. Maybe worth adding a small note in the README that this limitation exists and that Piranha JS is fairly experimental right now. 

Cross-file refactoring would be a lot more than most (all?) other versions of Piranha do, though, so that'd be cool! 

Medium term, my only worry here is people running this and it breaking their code because it removed a function called from another file. Specially annoying for JS, I think, since there is a chance such breakage wouldn't be detected except at runtime/prod.

Reply:
Good point. Ideally, integration tests of the target codebase should catch this. It will be useful to discuss this point in the README (not as a limitation but by explaining the current features of deep cleaning and the boundaries).